### PR TITLE
tenant: fixes a (timing) issue whenever an agent is removed and re-added

### LINKED ISSUE #1380: cannot add previously removed agent
Is your issue a feature request? If so, please raise it as an [enhancement](https://github.com/keylime/enhancements)

# Environment
 * OS / version: Fedora 37
 * Processor architecture: x86_64
 * TPM Manufacturer: swtpm
 * Keylime version: upstream da652daeec

# Description 
I can add an agent successfully but after removing it another attempt to add it fails

## Expected behavior vs. actual behavior 
agent should be added successfully

## Steps to reproduce problem
```
# keylime_tenant -v 127.0.0.1 -t 127.0.0.1 -u d432fbb3-d2f1-4a97-9ef7-75bd81c00000 --verify --runtime-policy policy.json --include payload-script --cert default -c add
# keylime_tenant -v 127.0.0.1 -t 127.0.0.1 -u d432fbb3-d2f1-4a97-9ef7-75bd81c00000  -c delete
# keylime_tenant -v 127.0.0.1 -t 127.0.0.1 -u d432fbb3-d2f1-4a97-9ef7-75bd81c00000 --verify --runtime-policy policy.json --include payload-script --cert default -c add
```

# Relevant logs
```
INFO:keylime.config:Reading configuration from ['/etc/keylime/logging.conf']
2023-05-16 14:15:20.579 - keylime.keylime.cmd_exec - DEBUG - Executing command "tpm2_startup --version"
2023-05-16 14:15:20.583 - keylime.tpm - INFO - TPM2-TOOLS Version: 5.4
2023-05-16 14:15:20.584 - keylime.tenant - WARNING - The configuration upgrade templates path /usr/share/keylime/templates does not exist
2023-05-16 14:15:20.585 - keylime.config - INFO - Reading configuration from ['/etc/keylime/tenant.conf']
2023-05-16 14:15:20.585 - keylime.tenant - INFO - Setting up client TLS...
2023-05-16 14:15:20.585 - keylime.tenant - INFO - Using default client_cert option for tenant
2023-05-16 14:15:20.585 - keylime.tenant - INFO - Using default client_key option for tenant
2023-05-16 14:15:20.585 - keylime.tenant - INFO - No value provided in client_key_password option for tenant, assuming the key is unencrypted
2023-05-16 14:15:20.600 - keylime.tenant - INFO - TLS is enabled.
2023-05-16 14:15:20.602 - urllib3.connectionpool - DEBUG - Starting new HTTPS connection (1): 127.0.0.1:8891
2023-05-16 14:15:20.621 - urllib3.connectionpool - DEBUG - https://127.0.0.1:8891/ "GET /v2.1/agents/d432fbb3-d2f1-4a97-9ef7-75bd81c00000 HTTP/1.1" 200 None
2023-05-16 14:15:20.636 - urllib3.connectionpool - DEBUG - Starting new HTTPS connection (1): 127.0.0.1:9002
2023-05-16 14:15:20.653 - urllib3.connectionpool - DEBUG - https://127.0.0.1:9002/ "GET /version HTTP/1.1" 200 69
2023-05-16 14:15:20.654 - keylime.cli.policies - INFO - TPM PCR Mask from policy is 0x0
2023-05-16 14:15:20.654 - keylime.ima - DEBUG - Loading runtime policy from policy.json
2023-05-16 14:15:20.654 - keylime.ima - DEBUG - Loaded runtime policy from policy.json with checksum b86c70739f11422ee4ae349f8c40c7ba7803a0b5d0bc448df850166e9ecd543c
2023-05-16 14:15:20.654 - keylime.config - INFO - Reading configuration from ['/etc/keylime/ca.conf']
2023-05-16 14:15:20.655 - keylime.ca-util - WARNING - The 'default' password option from CA configuration file cannot be used with keylime CLI (keylime_tenant or keylime_ca)
Please enter the password to decrypt your keystore: 
2023-05-16 14:15:23.027 - keylime.ca-util - INFO - Creating cert package for d432fbb3-d2f1-4a97-9ef7-75bd81c00000 in d432fbb3-d2f1-4a97-9ef7-75bd81c00000-pkg.zip
2023-05-16 14:15:23.049 - keylime.ca-util - INFO - Creating cert package for RevocationNotifier in RevocationNotifier-pkg.zip
2023-05-16 14:15:23.051 - urllib3.connectionpool - DEBUG - Starting new HTTPS connection (1): 127.0.0.1:9002
2023-05-16 14:15:23.139 - urllib3.connectionpool - DEBUG - https://127.0.0.1:9002/ "GET /v2.0/quotes/identity?nonce=aJ9JHgTQl2R2YwPgk8PB HTTP/1.1" 200 2009
{'code': 200, 'status': 'Success', 'results': {'quote': 'r/1RDR4AYACIAC8bQHs/PRBTiBTzFUBc4XHN6NNGN57Eoax7lZfm2hfKHABRhSjlKSGdUUWwyUjJZd1BnazhQQgAAAAAAANd/AAAAAgAAAAABIBkQIwAWNjYAAAABAAsDAAABACAtddKXP2f5j0Prvy4BNZuVZRX2xFG2fH4LTgtKSJYZTA==:ABQACwEAHcc7c2yQ0bMHLdfz8g3j1ilGx13Ado5NAjONNaJf03yItchtfbDaidFaaVyj7yr+9qOIoRVU2DPilqnxTmdN+hwpheAwKADEljMCOmlFPjcU8S0yE3uM9F1FZ0d2R0jdkMWCYwUBBqPUjWViHm5rDA2t1eS6FEUlFOwM+Hnhd6rSZp2TxlzJR5OF6vx/81bNMVYBSa+mrws++8HTo95+qVoXWMi9QjrkkG4gO0ficHSdLMvzaYGpXdAAhzbHEKPAB+NuOq5PNfYM50dRUE7PvnoW/mImcTVk6VBbiVT1f61InpytEpbuNgBxBiC7MAvkTe3CRY/eN8KYXi5xydOEIg==:AQAAAAsAAwAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAEAAAAgAOTtYzActvRdPDUuwU0fNAhz4FbjtNR0efv0Scz0Rx6FAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=', 'hash_alg': 'sha256', 'enc_alg': 'rsa', 'sign_alg': 'rsassa', 'pubkey': '-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAv46hndi2wDeXio3GBkCa\nrNUXmlpHg4FmQXok3upTFe7X2X4Vrft09av/a+RYj4I3Y5qmlnMsy+xWVLXAw5t3\nplKrpvwt10LKXtV29AyLq6qSSA82vuZHbAe642WdFbFDY3k2gmeoIYTriBMrVbMr\nGL4v3OViXeNwv02PwPLfJWf7Uf+OqJd8wu8foHzVn16PCD8W5ANoGSYLLgJvIiCO\nJZX/z5sRWtBV1dhVn0WCJuzH7ov3s1wD1uzFxOlLzeTLotdqijDfzpmUye74TOKt\n8cmRFFibqBor6FbApQ8W///CvfKiILldom8Mw1gBO3UFHauCzeg02qpBhT8LoWNU\nwwIDAQAB\n-----END PUBLIC KEY-----\n'}}
2023-05-16 14:15:23.140 - keylime.tenant - DEBUG - Agent_quote received quote: r/1RDR4AYACIAC8bQHs/PRBTiBTzFUBc4XHN6NNGN57Eoax7lZfm2hfKHABRhSjlKSGdUUWwyUjJZd1BnazhQQgAAAAAAANd/AAAAAgAAAAABIBkQIwAWNjYAAAABAAsDAAABACAtddKXP2f5j0Prvy4BNZuVZRX2xFG2fH4LTgtKSJYZTA==:ABQACwEAHcc7c2yQ0bMHLdfz8g3j1ilGx13Ado5NAjONNaJf03yItchtfbDaidFaaVyj7yr+9qOIoRVU2DPilqnxTmdN+hwpheAwKADEljMCOmlFPjcU8S0yE3uM9F1FZ0d2R0jdkMWCYwUBBqPUjWViHm5rDA2t1eS6FEUlFOwM+Hnhd6rSZp2TxlzJR5OF6vx/81bNMVYBSa+mrws++8HTo95+qVoXWMi9QjrkkG4gO0ficHSdLMvzaYGpXdAAhzbHEKPAB+NuOq5PNfYM50dRUE7PvnoW/mImcTVk6VBbiVT1f61InpytEpbuNgBxBiC7MAvkTe3CRY/eN8KYXi5xydOEIg==:AQAAAAsAAwAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAEAAAAgAOTtYzActvRdPDUuwU0fNAhz4FbjtNR0efv0Scz0Rx6FAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
2023-05-16 14:15:23.140 - keylime.tenant - DEBUG - Agent_quote received public key: -----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAv46hndi2wDeXio3GBkCa
rNUXmlpHg4FmQXok3upTFe7X2X4Vrft09av/a+RYj4I3Y5qmlnMsy+xWVLXAw5t3
plKrpvwt10LKXtV29AyLq6qSSA82vuZHbAe642WdFbFDY3k2gmeoIYTriBMrVbMr
GL4v3OViXeNwv02PwPLfJWf7Uf+OqJd8wu8foHzVn16PCD8W5ANoGSYLLgJvIiCO
JZX/z5sRWtBV1dhVn0WCJuzH7ov3s1wD1uzFxOlLzeTLotdqijDfzpmUye74TOKt
8cmRFFibqBor6FbApQ8W///CvfKiILldom8Mw1gBO3UFHauCzeg02qpBhT8LoWNU
wwIDAQAB
-----END PUBLIC KEY-----

2023-05-16 14:15:23.140 - keylime.tenant - DEBUG - Agent_quote received hash algorithm: sha256
2023-05-16 14:15:23.141 - keylime.tenant - DEBUG - Agent_quote received encryption algorithm: rsa
2023-05-16 14:15:23.141 - keylime.tenant - DEBUG - Agent_quote received signing algorithm: rsassa
2023-05-16 14:15:23.142 - keylime.tenant - WARNING - DANGER: EK cert checking is disabled and no additional checks on EKs have been specified with ek_check_script option. Keylime is not secure!!
2023-05-16 14:15:23.142 - keylime.tenant - INFO - Quote from 127.0.0.1 validated
2023-05-16 14:15:23.142 - keylime.tenant - DEBUG - b64_encrypted_u: Omy0qgo0O9A7NVJu3pH4yku1Bft42Eq5E2EoTjwHRg6diQ2t3i8Yd3Bjwd9LcyZJQPY/dhPDXUTNc6sTXH8hfXK1+wIWjXUF7g570bzu2TLlzCe5c5woeGAvCx8yf1HqCosfQb4SX+B9XiF/ZFEtOZzKPMCmREJg0DZtbkapl5tqJNYgsYyutzy8SDcEsauU1n4jFu0UkyexHy8MBxESheyLFbXp98at2gtSNHPNDUiq8RRlnE9jUxnhd2RzhLD0NluTOf6NGO3zTmtXxEy/mCYFW5Cy3II+JUmXxc4b9+GFCI+LjHvNVYfX5Sfrx+w8sujCtAQCV86NrzeqKPbtaw==
2023-05-16 14:15:23.143 - urllib3.connectionpool - DEBUG - Starting new HTTPS connection (1): 127.0.0.1:9002
2023-05-16 14:15:23.159 - urllib3.connectionpool - DEBUG - https://127.0.0.1:9002/ "POST /v2.0/keys/ukey HTTP/1.1" 200 46
2023-05-16 14:15:23.160 - keylime.tenant - DEBUG - b64_v: zJEaAACsFWX+oVUmaQHe2NGz+qB01m2mspctv1R9/24=
2023-05-16 14:15:23.161 - urllib3.connectionpool - DEBUG - Starting new HTTPS connection (1): 127.0.0.1:8881
2023-05-16 14:15:23.206 - urllib3.connectionpool - DEBUG - https://127.0.0.1:8881/ "POST /v2.1/agents/d432fbb3-d2f1-4a97-9ef7-75bd81c00000 HTTP/1.1" 200 49
2023-05-16 14:15:23.208 - urllib3.connectionpool - DEBUG - Starting new HTTPS connection (1): 127.0.0.1:8881
2023-05-16 14:15:23.236 - urllib3.connectionpool - DEBUG - https://127.0.0.1:8881/ "GET /v2.1/agents/d432fbb3-d2f1-4a97-9ef7-75bd81c00000 HTTP/1.1" 200 784
2023-05-16 14:15:23.236 - keylime.tenant - INFO - Agent Info:
{"d432fbb3-d2f1-4a97-9ef7-75bd81c00000": {"operational_state": "Start", "v": "zJEaAACsFWX+oVUmaQHe2NGz+qB01m2mspctv1R9/24=", "ip": "127.0.0.1", "port": 9002, "tpm_policy": "{\"mask\": \"0x400\"}", "meta_data": "{\"cert_serial\": 2, \"subject\": \"OU=53,O=MITLL,L=Lexington,ST=MA,CN=d432fbb3-d2f1-4a97-9ef7-75bd81c00000,C=US\"}", "has_mb_refstate": 0, "has_runtime_policy": 1, "accept_tpm_hash_algs": ["sha512", "sha384", "sha256"], "accept_tpm_encryption_algs": ["ecc", "rsa"], "accept_tpm_signing_algs": ["ecschnorr", "rsassa"], "hash_alg": "", "enc_alg": "", "sign_alg": "", "verifier_id": "default", "verifier_ip": "127.0.0.1", "verifier_port": 8881, "severity_level": null, "last_event_id": null, "attestation_count": 0, "last_received_quote": 0, "last_successful_attestation": 0}}
2023-05-16 14:15:23.236 - keylime.tenant - INFO - Agent ID d432fbb3-d2f1-4a97-9ef7-75bd81c00000 added to the Verifier after 0 tries
2023-05-16 14:15:23.237 - urllib3.connectionpool - DEBUG - Starting new HTTPS connection (1): 127.0.0.1:9002
2023-05-16 14:15:23.256 - urllib3.connectionpool - DEBUG - https://127.0.0.1:9002/ "GET /v2.0/keys/verify?challenge=UbjTVkdruKf6YTVtjpNC HTTP/1.1" 200 149
2023-05-16 14:15:23.257 - keylime.tenant - ERROR - Key derivation failed
2023-05-16 14:15:23.257 - urllib3.connectionpool - DEBUG - Starting new HTTPS connection (1): 127.0.0.1:8881
2023-05-16 14:15:23.280 - urllib3.connectionpool - DEBUG - https://127.0.0.1:8881/ "PUT /v2.1/agents/d432fbb3-d2f1-4a97-9ef7-75bd81c00000/stop HTTP/1.1" 200 49
2023-05-16 14:15:23.281 - keylime.tenant - INFO - Agent d432fbb3-d2f1-4a97-9ef7-75bd81c00000 stopped
```

### PR DESCRIPTION
**Pull request by maugustosilva**
The following timing issue has been detected on the `agent` (thanks @ansasaki for the help debugging): `agent` does not replace K until both the new U and V are received. The `tenant` tries to verify (due to the cli option `--verify`) that the `agent` has the new K before it receives the V part. For this request, the `agent` will use the old K, resulting in failure on `tenant` side.

This minimalistic fix implements a retry (actually it just re-uses the retry already existing on the code), just to give the `verifier` enough time to send the new `V`.

Fixes #1380

### CODE REVIEW

File: keylime/tenant.py [hunk 1]
```diff
@@ -1171,22 +1171,28 @@ def do_verify(self) -> None:
                     continue
                 self.do_cvstop()
                 raise e
+
+            mac = ""
+            ex_mac = crypto.do_hmac(self.K, challenge)
+
             if response.status_code == 200:
```
Comment (stefanberger) [inline keylime/tenant.py:1178]:
```
mac = response_json.get("results",{}).get("hmac")
if not mac:
    logger.critical(...
```

File: keylime/tenant.py [hunk 2]
```diff
@@ -1171,22 +1171,28 @@ def do_verify(self) -> None:
...
-                ex_mac = crypto.do_hmac(self.K, challenge)
-
                 if mac == ex_mac:
-                    logger.info("Key derivation successful")
-                else:
-                    logger.error("Key derivation failed")
-                    self.do_cvstop()
-            else:
-                keylime_logging.log_http_response(logger, logging.ERROR, response_json)
+                    logger.info("Agent on %s with port %s successfully derived key", self.agent_ip, self.agent_port)
```
Comment (stefanberger) [inline keylime/tenant.py:1191]: I wonder whether this is a mere hmac comparison rather than a key derivation ...

File: keylime/tenant.py (hunk 1 above)
Comment (maugustosilva) [reply keylime/tenant.py:1178]: In the original code, a response code other than `200` triggered the execution of `keylime_logging.log_http_response`, and I just elected to preserve it. Trying to make it as minimalistic as possible... this (`tenant`) code will require a more extensive overhaul in the future (we really need to revisit the whole protocol), and for now I just want to make sure we stop getting random (timing) errors when PRs are open.

File: keylime/tenant.py (hunk 2 above)
Comment (maugustosilva) [reply keylime/tenant.py:1191]: Well, it is a matter of semantics, I would say.  We are comparing `hmacs` as an evidence of possession of `K` by both parties, no? If you insist, I can change the message, but again, my goal was "the minimal set of changes which prevents errors and let others keep opening PRs"

Comment (stefanberger) [reply keylime/tenant.py:1191]: Nah, leave the message. I am just wondering (out loud) what this is ...

**Review by stefanberger — APPROVED** [review]

**Review by aplanas — APPROVED** [review]

**Pull request merged**